keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/ArtifactUsedConstraintEvaluator.kt

```diff
@@ -0,0 +1,32 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.core.api.ArtifactUsedConstraint
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+@Component
+class ArtifactUsedConstraintEvaluator(
+  override val eventPublisher: ApplicationEventPublisher
+) : ConstraintEvaluator<ArtifactUsedConstraint> {
+  override val supportedType = SupportedConstraintType<ArtifactUsedConstraint>("artifact-used")
+  override fun isImplicit() = true
+
+  override fun canPromote(artifact: DeliveryArtifact, version: String, deliveryConfig: DeliveryConfig, targetEnvironment: Environment): Boolean {
+    val allowedReferences = mutableSetOf<String>()
```

Comment:
If you want to improve your functional programming score, I think you can simplify the body of this function to:

```kotlin
return deliveryConfig
      .environments
      .firstOrNull { it.name == targetEnvironment.name }
      ?.resources
      ?.mapNotNull { it.findAssociatedArtifact(deliveryConfig) }
      ?.any { it.reference == artifact.reference }
      ?: false
```

Reply:
My personal preference is to not do anything too functional. That is maybe controversial, but I personally find functional programming hard to read and I'm bad at it. Also, the functional style is really different from the rest of our code, so I find it hard to debug when switching back and forth